### PR TITLE
Arbitrary tau sam tov

### LINKED ISSUE #247: Allow for arbitrary tau selection in MSD calculations.
**What feature would you like to see added?**
It would be helpful to have the ability to select not only a delta tau of 1 when performing an MSD calculation. That is, to compute the MSD at steps 0, 2, 4, ... rather than at each time-step.

**Describe how this feature would improve the code**
It would become more flexible.

**Describe how this feature would work/possible implementations**
The user should be able to parse a taus argument which can be in integer (points spaced by n) or a list specificity exactly which points to take.

Code-wise the `msd_operation` method will need to be changed. It should suffice to change the 
```python
 if square:
      return tf.math.squared_difference(ensemble, ensemble[:, None, 0])
```
portion to 

```python
 if square:
      return tf.math.squared_difference(ensemble[:, indices], ensemble[:, None, 0])
```

### PR DESCRIPTION
**Pull request by SamTov**
Fix #247 

#### Summary of additions and changes

* Allo for arbitrary tau selection in the Einstein diffusion coefficients.

#### How to test this pull request
Run some einstein diffusion as follows.
```
experiment.run_computation.EinsteinDiffusionCoefficients(tau_values=np.s_[::3])  # take every third
experiment.run_computation.EinsteinDiffusionCoefficients(tau_values=11)  # take 11 points evenly spaced
experiment.run_computation.EinsteinDiffusionCoefficients(tau_values=[1, 12, 13, 18, 22, 46])  # take the steps in the list.
```

### CODE REVIEW

File: mdsuite/calculators/calculator.py [hunk 1]
```diff
@@ -364,10 +378,26 @@ def _msd_operation(ensemble: tf.Tensor, square: bool = True):
                 Mean square displacement.
         """
         if square:
-            return tf.math.squared_difference(ensemble, ensemble[:, None, 0])
+            return tf.math.squared_difference(np.array(ensemble)[:, self.tau_values], ensemble[:, None, 0])
```
Comment (PythonFZ) [inline mdsuite/calculators/calculator.py:381]: I'd suggest for better performance that we don't use numpy here.
If we need the indexing I'd suggest using `tf.gather` or `tf.gather_nd` for this

File: mdsuite/calculators/calculator.py [hunk 2]
```diff
@@ -364,10 +378,26 @@ def _msd_operation(ensemble: tf.Tensor, square: bool = True):
...
+        -------
+        Updates the class state.
+        """
+        if type(self.tau_values) is int:
+            self.data_resolution = self.tau_values
+            self.tau_values = np.linspace(0, self.data_range - 1, int(self.tau_values), dtype=int)
+        if type(self.tau_values) is list:
+            self.data_resolution = len(self.tau_values)
+        if type(self.tau_values) is slice:
+            self.data_resolution = len(np.linspace(0, self.data_range, self.data_range)[self.tau_values])
```
Comment (PythonFZ) [inline mdsuite/calculators/calculator.py:400]: Isn't this just `len(np.arange(self.data_range)[self.tau_values])` 
I don't know the type slice, but I guess `len(self.tau_values)` does not work?

File: mdsuite/calculators/calculator.py [hunk 3]
```diff
@@ -452,11 +482,14 @@ def _save_data(self, name: str, data: pd.DataFrame):
                 db[self.database_group].create_dataset(title, data=data, dtype=float)
         """
 
-    def _plot_fig(self, fig: matplotlib.figure.Figure, ax: Axes,
+    def _plot_fig(self,
+                  fig: matplotlib.figure.Figure,
+                  ax: Axes,
```
Comment (PythonFZ) [inline mdsuite/calculators/calculator.py:488]: Why import `Axes` but not `Figure` ?

**Review by PythonFZ — CHANGES_REQUESTED** [review]

File: mdsuite/calculators/calculator.py (hunk 2 above)
Comment (SamTov) [reply mdsuite/calculators/calculator.py:400]: no it's okay how it is. The data resolution is just a scaling factor to re-scale the time arrays or msd arrays. 

```python
time = np.linspace(0, self.data-range, self.data-resolution)
```

File: mdsuite/calculators/calculator.py (hunk 1 above)
Comment (SamTov) [reply mdsuite/calculators/calculator.py:381]: Yeah this is okay but of the three outcomes only two require the gather, the np.s_ object works without numpy

File: mdsuite/calculators/calculator.py (hunk 3 above)
Comment (SamTov) [reply mdsuite/calculators/calculator.py:488]: I didn't add this, i just reformatted it.

File: mdsuite/calculators/calculator.py (hunk 2 above)
Comment (PythonFZ) [reply mdsuite/calculators/calculator.py:400]: Okay, I checked and

```py
 np.linspace(0, x, x - 1) == np.arange(x)
```

So I guess this is resolved then

File: mdsuite/calculators/calculator.py (hunk 1 above)
Comment (SamTov) [reply mdsuite/calculators/calculator.py:381]: I will try to parse all the inputs as np.s_ objects, it could work.

Comment (SamTov) [reply mdsuite/calculators/calculator.py:381]: Or can try to parse the np.s_ object to the gather. The other approach failed.

Comment (SamTov) [reply mdsuite/calculators/calculator.py:381]: Okay so I can just make the slices arrays and then gather works fine.

**Review by SamTov — COMMENTED** [review]
Updated all requests to completely use TensorFlow.  Code has been tested for all slice types and on a large system of emimF

**Review by PythonFZ — APPROVED** [review]

**Pull request merged**